File: android/src/main/kotlin/com/example/beauty_camera_plugin/BeautyCameraPlugin.kt
package com.example.beauty_camera_plugin

import android.app.Activity
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.StandardMessageCodec

/** BeautyCameraPlugin */
import android.view.Surface // Required for Surface.ROTATION_* constants

class BeautyCameraPlugin: FlutterPlugin, MethodCallHandler, ActivityAware {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel
  private lateinit var orientationEventChannel: EventChannel
  private var activity: Activity? = null
  
  // Camera manager instance
  private var cameraManager: CameraManager? = null
  private var currentPlatformView: CameraPlatformView? = null



  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "com.example/beauty_camera_plugin")
    channel.setMethodCallHandler(this)

    // Set up orientation event channel
    orientationEventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "com.example/beauty_camera_plugin/orientation")

    // Register platform view factory for camera preview
    flutterPluginBinding
      .platformViewRegistry
      .registerViewFactory(
        "com.example/camera_preview_view", 
        CameraPlatformViewFactory(StandardMessageCodec.INSTANCE, this)
      )
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    val manager = cameraManager
    if (manager == null && call.method != "getPlatformVersion" && call.method != "initializeCamera") {
      result.error("NOT_INITIALIZED", "Camera manager not initialized", null)
      return
    }

    when (call.method) {
      "getPlatformVersion" -> {
        result.success("Android ${android.os.Build.VERSION.RELEASE}")
      }        "initializeCamera" -> {
        activity?.let { 
          cameraManager = CameraManager(it)
          cameraManager?.initialize { success, error ->
            if (success) {
              // Set up orientation stream
              val handler = cameraManager?.orientationStreamHandler
              android.util.Log.d("BeautyCameraPlugin", "Setting OrientationStreamHandler: $handler")
              orientationEventChannel.setStreamHandler(handler)
              
              // If platformView is already registered, attach the manager
              currentPlatformView?.let { pv ->
                cameraManager?.let { cm ->
                  pv.attachCameraManager(cm)
                }
              }
              result.success(null)
            } else {
              android.util.Log.e("BeautyCameraPlugin", "CameraManager initialization failed: $error")
              result.error("INITIALIZATION_FAILED", error, null)
            }
          }
        } ?: result.error("NO_ACTIVITY", "Activity not available", null)
      }
      
      "takePicture" -> {
        manager!!.takePicture { filePath, error ->
          if (filePath != null) {
            result.success(filePath)
          } else {
            result.error("CAPTURE_FAILED", error, null)
          }
        }
      }
      
      "setFilter" -> {
        val filterType = call.argument<String>("filterType")
        if (filterType != null) {
          manager!!.setFilter(filterType)
          result.success(null)
        } else {
          result.error("INVALID_ARGUMENT", "Filter type is required", null)
        }
      }
      
      "setFilterIntensity" -> {
        val intensity = call.argument<Double>("intensity")
        if (intensity != null) {
          manager!!.setFilterIntensity(intensity.toFloat())
          result.success(null)
        } else {
          result.error("INVALID_ARGUMENT", "Intensity is required", null)
        }
      }
      
      "switchCamera" -> {
        manager!!.switchCamera()
        result.success(null)
      }
      
      "toggleFlash" -> {
        val flashMode = manager!!.toggleFlash()
        result.success(flashMode)
      }
      
      "setZoom" -> {
        val zoom = call.argument<Double>("zoom")
        if (zoom != null) {
          manager!!.setZoom(zoom.toFloat())
          result.success(null)
        } else {
          result.error("INVALID_ARGUMENT", "Zoom value is required", null)
        }
      }
      
      "setExposure" -> {
        val exposure = call.argument<Double>("exposure")
        if (exposure != null) {
          manager!!.setExposure(exposure.toFloat())
          result.success(null)
        } else {
          result.error("INVALID_ARGUMENT", "Exposure value is required", null)
        }
      }
      
      "setFocusPoint" -> {
        val x = call.argument<Double>("x")
        val y = call.argument<Double>("y")
        if (x != null && y != null) {
          manager!!.setFocusPoint(x.toFloat(), y.toFloat())
          result.success(null)
        } else {
          result.error("INVALID_ARGUMENT", "Focus point x and y are required", null)
        }
      }
      
      "setAutoFocus" -> {
        val enabled = call.argument<Boolean>("enabled")
        if (enabled != null) {
          manager!!.setAutoFocus(enabled)
          result.success(null)
        } else {
          result.error("INVALID_ARGUMENT", "Auto focus enabled flag is required", null)
        }
      }
      
      "dispose" -> {
        manager!!.dispose()
        cameraManager = null
        result.success(null)
      }
      else -> {
        result.notImplemented()
      }
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
    orientationEventChannel.setStreamHandler(null)
    cameraManager?.dispose()
    cameraManager = null
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onDetachedFromActivityForConfigChanges() {
    activity = null
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onDetachedFromActivity() {
    activity = null
    cameraManager?.dispose()
    cameraManager = null
  }

  fun getCameraManager(): CameraManager? {
    return cameraManager
  }

  fun registerPlatformView(view: CameraPlatformView?) {
    this.currentPlatformView = view
    // If cameraManager is already initialized and a new view is registered, set it.
    if (view != null && cameraManager != null) {
        view.attachCameraManager(cameraManager!!)
    }
  }
}

File: ios/Classes/BeautyCameraPlugin.swift
import Flutter
import UIKit
import AVFoundation

public class BeautyCameraPlugin: NSObject, FlutterPlugin {
    static let channelName = "com.example/beauty_camera_plugin"
    static let orientationChannelName = "com.example/beauty_camera_plugin/orientation"
    static let platformViewTypeId = "com.example/camera_preview_view"
    
    private var cameraManager: CameraManager?
    private var orientationStreamHandler: OrientationStreamHandler?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: channelName, binaryMessenger: registrar.messenger())
        let instance = BeautyCameraPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        
        // Register orientation event channel
        let orientationChannel = FlutterEventChannel(name: orientationChannelName, binaryMessenger: registrar.messenger())
        instance.orientationStreamHandler = OrientationStreamHandler()
        orientationChannel.setStreamHandler(instance.orientationStreamHandler)
        
        // Register platform view for camera preview
        registrar.register(
            CameraPlatformViewFactory(messenger: registrar.messenger(), plugin: instance),
            withId: platformViewTypeId
        )
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)
            
        case "initializeCamera":
            initializeCamera(result: result)
            
        case "takePicture":
            takePicture(result: result)
            
        case "setFilter":
            if let args = call.arguments as? [String: Any],
               let filterType = args["filterType"] as? String {
                setFilter(filterType: filterType, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Filter type is required", details: nil))
            }
            
        case "setFilterIntensity":
            if let args = call.arguments as? [String: Any],
               let intensity = args["intensity"] as? Double {
                setFilterIntensity(intensity: Float(intensity), result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Intensity is required", details: nil))
            }
            
        case "switchCamera":
            switchCamera(result: result)
            
        case "toggleFlash":
            toggleFlash(result: result)
            
        case "setZoom":
            if let args = call.arguments as? [String: Any],
               let zoom = args["zoom"] as? Double {
                setZoom(zoom: Float(zoom), result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Zoom level is required", details: nil))
            }
            
        case "setExposure":
            if let args = call.arguments as? [String: Any],
               let exposure = args["exposure"] as? Double {
                setExposure(exposure: Float(exposure), result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Exposure value is required", details: nil))
            }
            
        case "setFocusPoint":
            if let args = call.arguments as? [String: Any],
               let x = args["x"] as? Double,
               let y = args["y"] as? Double {
                setFocusPoint(x: Float(x), y: Float(y), result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Focus point x and y are required", details: nil))
            }
            
        case "setAutoFocus":
            if let args = call.arguments as? [String: Any],
               let enabled = args["enabled"] as? Bool {
                setAutoFocus(enabled: enabled, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Auto focus enabled flag is required", details: nil))
            }
            
        case "dispose":
            dispose(result: result)
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func initializeCamera(result: @escaping FlutterResult) {
        guard cameraManager == nil else {
            result(nil)
            return
        }
        
        cameraManager = CameraManager()
        cameraManager?.initializeCamera { [weak self] error in
            DispatchQueue.main.async {
                if let error = error {
                    result(FlutterError(code: "CAMERA_INIT_ERROR", message: error.localizedDescription, details: nil))
                } else {
                    self?.orientationStreamHandler?.startListening()
                    result(nil)
                }
            }
        }
    }
    
    private func takePicture(result: @escaping FlutterResult) {
        guard let cameraManager = cameraManager else {
            result(FlutterError(code: "CAMERA_NOT_INITIALIZED", message: "Camera not initialized", details: nil))
            return
        }
        
        cameraManager.takePicture { imagePath, error in
            DispatchQueue.main.async {
                if let error = error {
                    result(FlutterError(code: "CAPTURE_ERROR", message: error.localizedDescription, details: nil))
                } else {
                    result(imagePath)
                }
            }
        }
    }
    
    private func setFilter(filterType: String, result: @escaping FlutterResult) {
        guard let cameraManager = cameraManager else {
            result(FlutterError(code: "CAMERA_NOT_INITIALIZED", message: "Camera not initialized", details: nil))
            return
        }
        
        cameraManager.setFilter(filterType: filterType) { error in
            DispatchQueue.main.async {
                if let error = error {
                    result(FlutterError(code: "FILTER_ERROR", message: error.localizedDescription, details: nil))
                } else {
                    result(nil)
                }
            }
        }
    }
    
    private func setFilterIntensity(intensity: Float, result: @escaping FlutterResult) {
        guard let cameraManager = cameraManager else {
            result(FlutterError(code: "CAMERA_NOT_INITIALIZED", message: "Camera not initialized", details: nil))
            return
        }
        
        cameraManager.setFilterIntensity(intensity: intensity) { error in
            DispatchQueue.main.async {
                if let error = error {
                    result(FlutterError(code: "FILTER_INTENSITY_ERROR", message: error.localizedDescription, details: nil))
                } else {
                    result(nil)
                }
            }
        }
    }
    
    private func switchCamera(result: @escaping FlutterResult) {
        guard let cameraManager = cameraManager else {
            result(FlutterError(code: "CAMERA_NOT_INITIALIZED", message: "Camera not initialized", details: nil))
            return
        }
        
        cameraManager.switchCamera { error in
            DispatchQueue.main.async {
                if let error = error {
                    result(FlutterError(code: "CAMERA_SWITCH_ERROR", message: error.localizedDescription, details: nil))
                } else {
                    result(nil)
                }
            }
        }
    }
    
    private func toggleFlash(result: @escaping FlutterResult) {
        guard let cameraManager = cameraManager else {
            result(FlutterError(code: "CAMERA_NOT_INITIALIZED", message: "Camera not initialized", details: nil))
            return
        }
        
        cameraManager.toggleFlash { error in
            DispatchQueue.main.async {
                if let error = error {
                    result(FlutterError(code: "FLASH_ERROR", message: error.localizedDescription, details: nil))
                } else {
                    result(nil)
                }
            }
        }
    }
    
    private func setZoom(zoom: Float, result: @escaping FlutterResult) {
        guard let cameraManager = cameraManager else {
            result(FlutterError(code: "CAMERA_NOT_INITIALIZED", message: "Camera not initialized", details: nil))
            return
        }
        
        cameraManager.setZoom(zoom: zoom) { error in
            DispatchQueue.main.async {
                if let error = error {
                    result(FlutterError(code: "ZOOM_ERROR", message: error.localizedDescription, details: nil))
                } else {
                    result(nil)
                }
            }
        }
    }
    
    private func setExposure(exposure: Float, result: @escaping FlutterResult) {
        guard let cameraManager = cameraManager else {
            result(FlutterError(code: "CAMERA_NOT_INITIALIZED", message: "Camera not initialized", details: nil))
            return
        }
        
        cameraManager.setExposure(exposure: exposure) { error in
            DispatchQueue.main.async {
                if let error = error {
                    result(FlutterError(code: "EXPOSURE_ERROR", message: error.localizedDescription, details: nil))
                } else {
                    result(nil)
                }
            }
        }
    }
    
    private func setFocusPoint(x: Float, y: Float, result: @escaping FlutterResult) {
        guard let cameraManager = cameraManager else {
            result(FlutterError(code: "CAMERA_NOT_INITIALIZED", message: "Camera not initialized", details: nil))
            return
        }
        
        cameraManager.setFocusPoint(x: x, y: y) { error in
            DispatchQueue.main.async {
                if let error = error {
                    result(FlutterError(code: "FOCUS_POINT_ERROR", message: error.localizedDescription, details: nil))
                } else {
                    result(nil)
                }
            }
        }
    }
    
    private func setAutoFocus(enabled: Bool, result: @escaping FlutterResult) {
        guard let cameraManager = cameraManager else {
            result(FlutterError(code: "CAMERA_NOT_INITIALIZED", message: "Camera not initialized", details: nil))
            return
        }
        
        cameraManager.setAutoFocus(enabled: enabled) { error in
            DispatchQueue.main.async {
                if let error = error {
                    result(FlutterError(code: "AUTO_FOCUS_ERROR", message: error.localizedDescription, details: nil))
                } else {
                    result(nil)
                }
            }
        }
    }
    
    private func dispose(result: @escaping FlutterResult) {
        orientationStreamHandler?.stopListening()
        cameraManager?.dispose()
        cameraManager = nil
        result(nil)
    }
    
    func getCameraManager() -> CameraManager? {
        return cameraManager
    }
}
